data_structures/binary_trees/segment_tree.c

```diff
@@ -0,0 +1,231 @@
+    size_t elem_size; /**< size of each data element */
+    size_t size;      /**< total size of array which segment tree represents*/
+    combine_function combine; /**< the function to be used to combine two node's
+                                 data to form parent's data */
+} segment_tree;
+
+/**
+ * @fn Builds a Segment tree
+ * It is assumed that leaves of tree already contains data.
+ * @param tree pointer to segment tree to be build
+ */
+void segment_tree_build(segment_tree *tree)
+{
+    size_t elem_size = tree->elem_size;
+    int index = (tree->size - 2);
+    size_t b, l, r;
+    char *ptr = (char *)tree->root;
+    for (; index >= 0; index--)
+    {
+        b = index * elem_size;
+        l = (2 * index + 1) * elem_size;
+        r = (2 * index + 2) * elem_size;
+        tree->combine(ptr + l, ptr + r, ptr + b);
+    }
+}
+
+/**
+ * @fn For point updates
+ * This function updates the element at given index and also updates segment
+ * tree accordingly
+ *
+ * @param tree pointer to segment tree
+ * @param index the index whose element is to be updated (0 based indexing used)
+ * @param val pointer to value that is to be replaced at given index
+ */
+void segment_tree_update(segment_tree *tree, size_t index, void *val)
+{
+    size_t elem_size = tree->elem_size;
+    index = index + tree->size - 1;
+    char *base = (char *)tree->root;
+    char *t = base + index * elem_size;
+    memcpy(t, val, elem_size);
+    while (index > 0)
+    {
+        index = ((index - 1) >> 1);
+        tree->combine(base + (2 * index + 1) * elem_size,
+                      base + (2 * index + 2) * elem_size,
+                      base + index * elem_size);
+    }
+}
+
+/**
+ * @fn Query the segment tree
+ * This function helps in range query of segment tree
+ * This function assumes that the given range is valid
+ * Performs the query in range [l,r]
+ * @param tree pointer to segment tree
+ * @param l the start of range
+ * @param r the end of range
+ * @param res the pointer to memory where result of query is stored
+ */
+void segment_tree_query(segment_tree *tree, long long l, long long r, void *res)
+{
+    size_t elem_size = tree->elem_size;
+    memcpy(res, tree->identity, elem_size);
+    elem_size = tree->elem_size;
+    char *root = (char *)tree->root;
+    l += tree->size - 1;
+    r += tree->size - 1;
+    while (l <= r)
+    {
+        if (!(l & 1))
+        {
+            tree->combine(res, root + l * elem_size, res);
+        }
+        if (r & 1)
+        {
+            tree->combine(res, root + r * elem_size, res);
+        }
+        r = (r >> 1) - 1;
+        l = (l >> 1);
+    }
+}
+
+/**
+ * @fn Initializes Segment Tree
+ * Accquires memory for segment tree
+ * and fill the leaves of segment tree with data from array
+ * @param arr the array data upon which segment tree is build
+ * @param elem_size size of each element in segment tree
+ * @param size total no of elements in array
```

Comment:
```suggestion
 * @param size total no of elements in array in bytes 
```
it think it is important to mention that. Otherwise, passing the number of array elements of type `int` would lead to incorrect results

Reply:
Sorry actually the description is correct size here is length of array my bad I must have used length or len instead of size I will correct it